naming/program/src/processor.rs

```diff
@@ -0,0 +1,217 @@
+            return Err(ProgramError::InvalidArgument);
+        }
+        if name_record_header.class != Pubkey::default()
+            && *name_update_signer.key != name_record_header.class
+        {
+            msg!("The given name class account is incorrect.");
+            return Err(ProgramError::InvalidArgument);
+        }
+        if name_record_header.class == Pubkey::default()
+            && *name_update_signer.key != name_record_header.owner
+        {
+            msg!("The given name owner account is incorrect.");
+            return Err(ProgramError::InvalidArgument);
+        }
+
+        write_data(name_account, &data, offset);
+
+        Ok(())
+    }
+
+    pub fn process_transfer(accounts: &[AccountInfo], new_owner: Pubkey) -> ProgramResult {
+        let accounts_iter = &mut accounts.iter();
+
+        let name_account = next_account_info(accounts_iter)?;
+        let name_owner = next_account_info(accounts_iter)?;
+        let name_class_opt = next_account_info(accounts_iter).ok();
+
+        let mut name_record_header =
+            NameRecordHeader::unpack_from_slice(&name_account.data.borrow())?;
+
+        // Verifications
+        if !name_owner.is_signer || name_record_header.owner != *name_owner.key {
+            msg!("The given name owner is incorrect or not a signer.");
+            return Err(ProgramError::InvalidArgument);
+        }
+        if name_record_header.class != Pubkey::default()
+            && (name_class_opt.is_none()
+                || name_record_header.class != *name_class_opt.unwrap().key
+                || !name_class_opt.unwrap().is_signer)
+        {
+            msg!("The given name class account is incorrect or not a signer.");
+            return Err(ProgramError::InvalidArgument);
+        }
+
+        name_record_header.owner = new_owner;
+        name_record_header
+            .pack_into_slice(&mut name_account.data.borrow_mut()[..NameRecordHeader::LEN]);
+
+        Ok(())
+    }
+
+    pub fn process_delete(accounts: &[AccountInfo]) -> ProgramResult {
+        let accounts_iter = &mut accounts.iter();
+
+        let name_account = next_account_info(accounts_iter)?;
+        let name_owner = next_account_info(accounts_iter)?;
+        let refund_target = next_account_info(accounts_iter)?;
+
+        let name_record_header = NameRecordHeader::unpack_from_slice(&name_account.data.borrow())?;
+
+        // Verifications
+        if !name_owner.is_signer || name_record_header.owner != *name_owner.key {
+            msg!("The given name owner is incorrect or not a signer.");
+            return Err(ProgramError::InvalidArgument);
+        }
+
+        // Close the account by transferring the rent sol
+        let source_amount: &mut u64 = &mut name_account.lamports.borrow_mut();
+        let dest_amount: &mut u64 = &mut refund_target.lamports.borrow_mut();
+        *dest_amount += *source_amount;
+        *source_amount = 0;
+
```

Comment:
Let's zero out the full account data here as well and serialize it back into `name_account`.  It's possible that the Delete instruction is called from another program, and that program tries to do something with `name_account` after it's deleted here